src/realmd/AuthSocket.cpp

```diff
@@ -575,10 +584,23 @@ bool AuthSocket::_HandleLogonProof()
         self->srp.HashSessionKey();
         self->srp.CalculateProof(self->_login);
 
+        bool pinResult = true;
+
+        if (self->m_promptPin && (lp->securityFlags & SECURITY_FLAG_PIN))
+            pinResult = false;
+
+        if (self->m_promptPin && (lp->securityFlags & SECURITY_FLAG_PIN) && !self->_token.empty())
+        {
+            auto pin = self->generateToken(self->_token.c_str());
+
+            if (pin != uint32(-1))
+                pinResult = self->VerifyPinData(pin, lp->pinData);
+        }
+
         ///- Check if SRP6 results match (password is correct), else send an error
-        if (!self->srp.Proof(lp->M1, 20))
+        if (!self->srp.Proof(lp->M1, 20) && pinResult)
```

Comment:
You flipped the flow of code. All code first checks password, then pin/authenticator. Unless there is a good reason for that, rewrite it so that proof is checked first and pin after

Reply:
should be possible, I'll change it